crates/task-impls/src/quorum_proposal.rs

```diff
@@ -0,0 +1,501 @@
+        network::{ConnectedNetwork, ConsensusIntentEvent},
+        node_implementation::{ConsensusTime, NodeImplementation, NodeType},
+    },
+    vote::HasViewNumber,
+};
+
+#[cfg(async_executor_impl = "async-std")]
+use async_std::task::JoinHandle;
+#[cfg(async_executor_impl = "tokio")]
+use tokio::task::JoinHandle;
+use tracing::{debug, error, instrument};
+
+use crate::{
+    consensus::CommitmentAndMetadata,
+    events::HotShotEvent,
+    helpers::{broadcast_event, cancel_task},
+};
+
+/// Validate a quorum proposal.
+#[allow(clippy::needless_pass_by_value)]
+fn validate_quorum_proposal<TYPES: NodeType>(
+    _quorum_proposal: Proposal<TYPES, QuorumProposal<TYPES>>,
+    _event_sender: Sender<Arc<HotShotEvent<TYPES>>>,
+) -> bool {
+    true
+}
+
+/// Validate a quorum proposal.
+#[allow(clippy::needless_pass_by_value)]
+fn validate_view_change_evidence<TYPES: NodeType>(
+    _certificate: ViewChangeEvidence<TYPES>,
+    _event_sender: Sender<Arc<HotShotEvent<TYPES>>>,
+) -> bool {
+    true
+}
+
+/// Validate a quorum proposal.
+#[allow(clippy::needless_pass_by_value)]
+fn validate_quorum_certificate<TYPES: NodeType>(
+    _certificate: QuorumCertificate<TYPES>,
+    _event_sender: Sender<Arc<HotShotEvent<TYPES>>>,
+) -> bool {
+    true
+}
+
+/// Validate a quorum proposal.
+#[allow(clippy::needless_pass_by_value)]
+fn validate_payload_and_metadata<TYPES: NodeType>(
+    _payload_commitment_and_metadata: CommitmentAndMetadata<<TYPES as NodeType>::BlockPayload>,
+    _event_sender: Sender<Arc<HotShotEvent<TYPES>>>,
+) -> bool {
+    true
+}
+
+/// Proposal dependency types. These types represent events that precipitate a proposal.
+#[derive(PartialEq)]
+enum ProposalDependency {
+    /// For the `SendPayloadCommitmentAndMetadata` event.
+    PayloadAndMetadata,
+
+    /// For the `ViewSyncFinalizeCertificate2Recv` event.
+    ViewSync,
+
+    /// For the `QuorumProposalRecv` event.
+    QuorumProposal,
+
+    /// For the `QCFormed` event.
+    QCFormed,
+}
+
+/// Handler for the proposal dependency
+struct ProposalDependencyHandle<TYPES: NodeType> {
+    /// The view number to propose for.
+    view_number: TYPES::Time,
+
+    /// The event sender.
+    sender: Sender<Arc<HotShotEvent<TYPES>>>,
+
+    /// Reference to consensus. The replica will require a write lock on this.
+    #[allow(dead_code)]
+    consensus: Arc<RwLock<Consensus<TYPES>>>,
+}
+impl<TYPES: NodeType> HandleDepOutput for ProposalDependencyHandle<TYPES> {
+    type Output = Vec<Arc<HotShotEvent<TYPES>>>;
+
+    #[allow(clippy::no_effect_underscore_binding)]
+    async fn handle_dep_result(self, res: Self::Output) {
+        let mut payload_commitment = None;
+        let mut commit_and_metadata: Option<CommitmentAndMetadata<TYPES::BlockPayload>> = None;
+        let mut _quorum_certificate = None;
+        let mut _timeout_certificate = None;
+        let mut _view_sync_finalize_cert = None;
+        for event in res {
+            match event.as_ref() {
```

Comment:
Keyao had defined some custom `*Validated` responses here but that won't work for me unfortunately so I instead propagate the event (see below comment) and then that ensures that this task is alive when these events arrive, so we should have no problem receiving them and proposing if all of the requisite information is available.

Reply:
Not sure I understand the difference. This implementation seems very similar to the `handle_dep_result` function for the quorum vote task.

I thought the difference was that we would not have `*Validated` events before creating dependencies and would use `mark_as_completed` to complete the dependency associated with the received event after creating the dependency. But looks like now we have both. Is the `mark_as_completed` calls only to replace the `validate_*` functions in the quorum vote task, rather than completely replace the `Validated*` events?

Reply:
This ended up changing a lot since this comment. In short, you approach ended up being the right one here. Sorry for the confusion.